# DS Automation Assignment

Using our prepared churn data from week 2:
- use pycaret to find an ML algorithm that performs best on the data
    - Choose a metric you think is best to use for finding the best model; by default, it is accuracy but it could be AUC, precision, recall, etc. The week 3 FTE has some information on these different metrics.
- save the model to disk
- create a Python script/file/module with a function that takes a pandas dataframe as an input and returns the probability of churn for each row in the dataframe
    - your Python file/function should print out the predictions for new data (new_churn_data.csv)
    - the true values for the new data are [1, 0, 0, 1, 0] if you're interested
- test your Python module and function with the new data, new_churn_data.csv
- write a short summary of the process and results at the end of this notebook
- upload this Jupyter Notebook and Python file to a Github repository, and turn in a link to the repository in the week 5 assignment dropbox

*Optional* challenges:
- return the probability of churn for each new prediction, and the percentile where that prediction is in the distribution of probability predictions from the training dataset (e.g. a high probability of churn like 0.78 might be at the 90th percentile)
- use other autoML packages, such as TPOT, H2O, MLBox, etc, and compare performance and features with pycaret
- create a class in your Python module to hold the functions that you created
- accept user input to specify a file using a tool such as Python's `input()` function, the `click` package for command-line arguments, or a GUI
- Use the unmodified churn data (new_unmodified_churn_data.csv) in your Python script. This will require adding the same preprocessing steps from week 2 since this data is like the original unmodified dataset from week 1.

In [1]:
import pandas as pd
df = pd.read_csv('prepped_churn_data.csv', index_col = 'customerID')
df

Unnamed: 0  tenure  PhoneService  Contract  PaymentMethod  \
customerID                                                              
7590-VHVEG           0       1             0         0              0   
5575-GNVDE           1      34             1         1              1   
3668-QPYBK           2       2             1         0              1   
7795-CFOCW           3      45             0         1              2   
9237-HQITU           4       2             1         0              0   
...                ...     ...           ...       ...            ...   
6840-RESVB        7038      24             1         1              1   
2234-XADUH        7039      72             1         1              3   
4801-JZAZL        7040      11             0         0              0   
8361-LTMKD        7041       4             1         0              1   
3186-AJIEK        7042      66             1         2              2   

            MonthlyCharges  TotalCharges  Churn  
customerID                                       
7590-VHVEG           29.85         29.85      0  
5575-GNVDE           56.95       1889.50      0  
3668-QPYBK           53.85        108.15      1  
7795-CFOCW           42.30       1840.75      0  
9237-HQITU           70.70        151.65      1  
...                    ...           ...    ...  
6840-RESVB           84.80       1990.50      0  
2234-XADUH          103.20       7362.90      0  
4801-JZAZL           29.60        346.45      0  
8361-LTMKD           74.40        306.60      1  
3186-AJIEK          105.65       6844.50      0  

[7043 rows x 8 columns]

In [2]:
!conda install -c conda-forge pycaret -y

Solving environment: done

# All requested packages already installed.



In [3]:
from pycaret.classification import setup, compare_models, predict_model, save_model, load_model

In [4]:
automl = setup(df, target = 'Churn')

In [5]:
best_model = compare_models()

In [6]:
best_model

GradientBoostingClassifier(ccp_alpha=0.0, criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='deviance', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=100,
                           n_iter_no_change=None, presort='deprecated',
                           random_state=8260, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

In [12]:
save_model(best_model, 'GBC')

Transformation Pipeline and Model Succesfully Saved


(Pipeline(memory=None,
          steps=[('dtypes',
                  DataTypes_Auto_infer(categorical_features=[],
                                       display_types=True, features_todrop=[],
                                       id_columns=['Unnamed: 0'],
                                       ml_usecase='classification',
                                       numerical_features=[], target='Churn',
                                       time_features=[])),
                 ('imputer',
                  Simple_Imputer(categorical_strategy='not_available',
                                 fill_value_categorical=None,
                                 fill_value_numerical=None,
                                 num...
                                             learning_rate=0.1, loss='deviance',
                                             max_depth=3, max_features=None,
                                             max_leaf_nodes=None,
                                             min_

In [13]:
from IPython.display import Code

In [14]:
Code('churn_predict.py')

import pandas as pd
from pycaret.classification import predict_model, load_model

def load_data(filepath):
    """
    Loads churn data into a DataFrame from a string filepath.
    """
    df = pd.read_csv(filepath, index_col='customerID')
    return df


def make_predictions(df):
    """
    Uses the pycaret best model to make predictions on data in the df dataframe.
    """
    model = load_model('GBC')
    predictions = predict_model(model, data=df)
    predictions.rename({'Label': 'Churn_prediction'}, axis=1, inplace=True)
    predictions['Churn_prediction'].replace({1: 'Churn', 0: 'No churn'},
                                            inplace=True)
    return predictions['Churn_prediction']


if __name__ == "__main__":
    df = load_data('new_churn_data.csv')
    predictions = make_predictions(df)
    print('predictions:')
    print(predictions)

In [15]:
%run churn_predict.py

Transformation Pipeline and Model Successfully Loaded
predictions:
customerID
9305-CKSKC    No churn
1452-KNGVK    No churn
6723-OKKJM    No churn
7832-POPKP    No churn
6348-TACGU    No churn
Name: Churn_prediction, dtype: object


In [23]:
predict_model(best_model, data=df)

tenure  PhoneService  Contract  PaymentMethod  MonthlyCharges  \
customerID                                                                  
9305-CKSKC      22             1         0              2           97.40   
1452-KNGVK       8             0         1              1           77.30   
6723-OKKJM      28             1         0              0           28.25   
7832-POPKP      62             1         0              2          101.70   
6348-TACGU      10             0         0              1           51.15   

            TotalCharges  charge_per_tenure  Label   Score  
customerID                                                  
9305-CKSKC        811.70          36.895455      0  0.5463  
1452-KNGVK       1701.95         212.743750      0  0.8743  
6723-OKKJM        250.90           8.960714      0  0.8430  
7832-POPKP       3106.56          50.105806      0  0.7441  
6348-TACGU       3440.97         344.097000      0  0.7559

# Summary

The gradient boosting classifier had the best performance on all metrics except recall and F1 (both of which were highest in the Naive Bayes model). However, the first five results when running churn_predict returned two false negatives (interestingly enough, a prior run of this same code returned the ADA boost classifier as the best model and got all five predictions correct on the %run output). The models are all under 80% accuracy, however, so some error is to be expected. The two false positives also happened to be the two rows with the lowest prediction probabilities (0.5463 and 0.7441, respectively).